File: BE/src/main/java/com/httmdt/orientalherbs/service/order/OrderService.java
package com.httmdt.orientalherbs.service.order;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.httmdt.orientalherbs.dto.order.OrderDetailResponse;
import com.httmdt.orientalherbs.dto.order.OrderListItemResponse;
import com.httmdt.orientalherbs.dto.order.OrderRequest;
import com.httmdt.orientalherbs.dto.order.OrderResponse;
import com.httmdt.orientalherbs.model.enums.OrderStatus;

public interface OrderService {
    // Create order
    OrderResponse createOrder(Long userId, OrderRequest request);

    // Get orders for user
    Page<OrderListItemResponse> getUserOrders(Long userId, Pageable pageable);
    Page<OrderListItemResponse> getUserOrdersByStatus(Long userId, OrderStatus status, Pageable pageable);

    // Get order detail for user
    OrderDetailResponse getUserOrderDetail(Long userId, Long orderId);

    // Get all orders (admin)
    Page<OrderListItemResponse> getAllOrders(Pageable pageable);
    Page<OrderListItemResponse> getOrdersByStatus(OrderStatus status, Pageable pageable);

    // Get order detail (admin)
    OrderDetailResponse getOrderDetail(Long orderId);

    // Update order status (admin)
    OrderDetailResponse updateOrderStatus(Long orderId, OrderStatus newStatus);

    // Cancel order (user)
    OrderDetailResponse cancelOrder(Long userId, Long orderId);
}

File: fe/src/services/order.service.ts
// fe/src/services/order.service.ts
import { http } from '@/utils/http';
import { OrderRequest, OrderResponse, OrderListItem, OrderDetail, OrderStatus } from '@/types/order.type';

// Response wrapper to preserve pagination info
export interface OrderListResponse {
  items: OrderListItem[];
  totalPages: number;
}

export const orderService = {
  // Create order
  createOrder: async (data: OrderRequest): Promise<OrderResponse> => {
    return await http<OrderResponse>('/orders', {
      method: 'POST',
      body: data,
    });
  },

  // Get user's orders with pagination
  getMyOrders: async (page: number = 0, size: number = 10): Promise<OrderListResponse> => {
    const response = await http<any>(`/orders?page=${page}&size=${size}`, {
      method: 'GET',
    });
    return {
      items: response.content || [],
      totalPages: response.totalPages || 0,
    };
  },

  // Get user's orders by status
  getMyOrdersByStatus: async (status: OrderStatus, page: number = 0, size: number = 10): Promise<OrderListResponse> => {
    const response = await http<any>(`/orders/status/${status}?page=${page}&size=${size}`, {
      method: 'GET',
    });
    return {
      items: response.content || [],
      totalPages: response.totalPages || 0,
    };
  },

  // Get order detail
  getOrderDetail: async (orderId: number): Promise<OrderDetail> => {
    return await http<OrderDetail>(`/orders/${orderId}`, {
      method: 'GET',
    });
  },

  // Cancel order
  cancelOrder: async (orderId: number): Promise<OrderDetail> => {
    return await http<OrderDetail>(`/orders/${orderId}/cancel`, {
      method: 'PUT',
    });
  },
};

